UnitystationLauncher/Services/ServerAuthenticationService.cs

```diff
@@ -0,0 +1,144 @@
+using System.Text;
+using System.Threading.Tasks;
+using System.Xml;
+using Newtonsoft.Json;
+using Org.BouncyCastle.Crypto;
+using Org.BouncyCastle.Crypto.Digests;
+using Org.BouncyCastle.Crypto.Encodings;
+using Org.BouncyCastle.Crypto.Engines;
+using Org.BouncyCastle.Crypto.Parameters;
+using Org.BouncyCastle.Math;
+using Org.BouncyCastle.OpenSsl;
+using UnitystationLauncher.Models;
+using UnitystationLauncher.Models.Api;
+
+namespace UnitystationLauncher.Services;
+
+public interface IServerAuthenticationService
+{
+    public Task<Server> GetServerInfoByIP(string IP);
+
+    public Task<Dictionary<string, string>> AuthenticateWithServer(string IP, Installation Installation);
+}
+
+public class ServerAuthenticationService : IServerAuthenticationService
+{
+    public ServerAuthenticationService(AuthService AuthService)
+    {
+        _AuthService = AuthService;
+    }
+
+
+    private readonly AuthService _AuthService;
+    private readonly HttpClient _httpClient = new HttpClient();
+
+    private readonly SHA512 SHA512 = SHA512.Create();
+
+    public async Task<Server> GetServerInfoByIP(string IP)
+    {
+        try
+        {
+            string Port = "7778";
+            string url = $"http://{IP}:{Port}/";
+
+            var response = await _httpClient.GetAsync(url);
+            response.EnsureSuccessStatusCode();
+
+            string content = await response.Content.ReadAsStringAsync();
+            var serverInfo = JsonConvert.DeserializeObject<Server>(content);
+
+            return serverInfo;
+        }
+        catch (Exception ex)
+        {
+            return null;
+        }
+    }
+
+
+    static AsymmetricKeyParameter ImportKeyFromPem(string pem)
+    {
+        using (var sr = new StringReader(pem))
+        {
+            var pr = new PemReader(sr);
+            return (AsymmetricKeyParameter)pr.ReadObject();
+        }
+    }
+
+
+    public async Task<Dictionary<string, string>> AuthenticateWithServer(string IP, Installation Installation)
+    {
+        var Info = await GetServerInfoByIP(IP);
+        var RSAEncrypt = new OaepEncoding(
+            new RsaEngine(),
+            new Sha256Digest()
+        );
+
+        RSAEncrypt.Init(true, ImportKeyFromPem(Info.ServerPublicKey)); // false = decrypt mode
+
+        byte[] sharedSecret = new byte[32]; // 256-bit key
+        RandomNumberGenerator.Fill(sharedSecret);
+
+        // Optional: convert to Base64 if you want to transmit/store it
+        string base64Secret = Convert.ToBase64String(sharedSecret);
+
+        var SHA512Check = Convert.ToBase64String(
+            SHA512.ComputeHash(
+                Encoding.UTF8.GetBytes(
+                    base64Secret
+                    + Info.ServerPublicKey
+                    + Installation.BuildVersion.ToString()
+                    + Installation.ForkName.ToString()
+                    + Installation.GoodFileVersion.ToString()
+                    + Info.ServerConnectionPublicKey)));
```

Comment:
Don't stuff the algorithm like this. While it's practically impossible for a hash collision to happen at these miniscule sizes, This extra data just has no business being hashed and will incur a (minor) performance penalty